pkg/storegateway/series_refs.go

```diff
@@ -444,3 +448,125 @@ func (s *deduplicatingSeriesChunkRefsSetIterator) Next() bool {
 	s.current = nextSet
 	return true
 }
+
+type limitingSeriesChunkRefsSetIterator struct {
+	from          seriesChunkRefsSetIterator
+	chunksLimiter ChunksLimiter
+	seriesLimiter SeriesLimiter
+
+	err        error
+	currentSet seriesChunkRefsSet
+}
+
+func (l *limitingSeriesChunkRefsSetIterator) Next() bool {
+	if l.err != nil {
+		return false
+	}
+
+	if !l.from.Next() {
+		l.err = l.from.Err()
+		return false
+	}
+
+	l.currentSet = l.from.At()
+	err := l.seriesLimiter.Reserve(uint64(l.currentSet.len()))
+	if err != nil {
+		l.err = errors.Wrap(err, "exceeded series limit")
+		return false
+	}
+
+	var totalChunks int
+	for _, s := range l.currentSet.series {
+		totalChunks += len(s.chunks)
+	}
+
+	err = l.chunksLimiter.Reserve(uint64(totalChunks))
+	if err != nil {
+		l.err = errors.Wrap(err, "exceeded chunks limit")
+		return false
+	}
+	return true
+}
+
+func (l *limitingSeriesChunkRefsSetIterator) At() seriesChunkRefsSet {
+	return l.currentSet
+}
+
+func (l *limitingSeriesChunkRefsSetIterator) Err() error {
+	return l.err
+}
+
+// TODO dimitarvdimitrov add a constructor and override minTime, maxTime when skipChunks==true; test already exists and is failing
+type inflatedSeriesChunkRefsSetIterator struct {
```

Comment:
This is done the real loading, so I would just call it for what it does.

```suggestion
type loadingSeriesChunkRefsSetIterator struct {
```